dhall-lsp-server/src/Dhall/LSP/Handlers.hs

```diff
@@ -2,59 +2,346 @@ module Dhall.LSP.Handlers where
+
+-- | A helper function to query haskell-lsp's VFS.
+readUri :: J.Uri -> HandlerM Text
+readUri uri = do
+  getVirtualFileFunc <- lift $ gets (LSP.getVirtualFileFunc . view lspFuncs)
+  asd <- liftIO $ getVirtualFileFunc uri
+  case asd of
+    Just (LSP.VirtualFile _ rope) -> return (Rope.toText rope)
+    Nothing -> fail $ "Could not find " <> show uri <> " in VFS."
+
+loadFile :: J.Uri -> HandlerM (Expr Src X)
+loadFile uri = do
+  txt <- readUri uri
+  fileIdentifier <- fileIdentifierFromUri uri
+  cache <- lift $ gets (view importCache)
+
+  expr <- case parse txt of
+    Right e -> return e
+    _ -> throwE (Error, "Failed to parse Dhall file.")
+
+  loaded <- liftIO $ load fileIdentifier expr cache
+  (cache', expr') <- case loaded of
+    Right x -> return x
+    _ -> throwE (Error, "Failed to resolve imports.")
+  -- Update cache. Don't cache current expression because it might not have been
+  -- written to disk yet (readUri reads from the VFS).
+  lift $ modify (set importCache cache')
+  return expr'
+
+-- helper
+fileIdentifierFromUri :: J.Uri -> HandlerM FileIdentifier
+fileIdentifierFromUri uri =
+  let mFileIdentifier = fmap fileIdentifierFromFilePath (J.uriToFilePath uri)
+                        <|> (do uri' <- (URI.parseURI . Text.unpack . J.getUri) uri
+                                fileIdentifierFromURI uri')
+  in case mFileIdentifier of
+    Just fileIdentifier -> return fileIdentifier
+    Nothing -> throwE (Error, J.getUri uri <> " is not a valid name for a dhall file.")
+
+-- helper
+rangeToJSON :: Range -> J.Range
+rangeToJSON (Range (x1,y1) (x2,y2)) = J.Range (J.Position x1 y1) (J.Position x2 y2)
+
+hoverExplain :: J.HoverRequest -> HandlerM ()
+hoverExplain request = do
+  let uri = request ^. J.params . J.textDocument . J.uri
+      (J.Position line col) = request ^. (J.params . J.position)
+  txt <- readUri uri
+  mError <- lift $ gets (Map.lookup uri . view errors)
+  let isHovered (Diagnosis _ (Just (Range left right)) _) =
+        left <= (line,col) && (line,col) <= right
+      isHovered _ = False
+
+      hoverFromDiagnosis (Diagnosis _ (Just (Range left right)) diagnosis) =
+        let _range = Just $ J.Range (uncurry J.Position left)
+                                    (uncurry J.Position right)
+            encodedDiag = URI.encode (Text.unpack diagnosis)
+            command = "[Explain error](dhall-explain:?"
+                        <> Text.pack encodedDiag <> " )"
+            _contents = J.List [J.PlainString command]
+        in Just J.Hover { .. }
+      hoverFromDiagnosis _ = Nothing
+
+      mHover = do err <- mError
+                  explanation <- explain txt err
+                  True <- return $ isHovered explanation
+                  hoverFromDiagnosis explanation
+  lspRespond LSP.RspHover request mHover
+
+hoverType :: J.HoverRequest -> HandlerM ()
+hoverType request = do
+  let uri = request ^. J.params . J.textDocument . J.uri
```

Comment:
Since lenses are first class, you can define a reusable composite lens from the request to the URI:

```haskell
requestUri = J.params . J.textDocument . J.uri
```

... and then you can reuse that for each handler:

```haskell
let uri = request ^. requestUri
```

Reply:
I tried that, but it doesn't work due to the way `haskell-lsp-types` is set up: we can't give a type to `requestUri` in general.

Reply:
Ah, I didn't realize that they were using chained multi-parameter type-classes.  Never mind, then